components/rustlib/CMakeLists.txt

```diff
@@ -0,0 +1,48 @@
+set(RUST_DEPS "clib")
+
+idf_component_register(
+    SRCS "placeholder.c"
+    INCLUDE_DIRS ""
+    PRIV_REQUIRES "${RUST_DEPS}"
+)
+
+set(CARGO_BUILD_TYPE "release")
+set(CARGO_BUILD_ARG "--release")
+if(CONFIG_IDF_TARGET_ARCH_RISCV)
+    set(CARGO_TARGET "riscv32i-unknown-none-elf")
+    set(CARGO_FEATURES_ARG "")
+elseif(CONFIG_IDF_TARGET_ARCH_XTENSA)
+    set(CARGO_TARGET "xtensa-esp32-none-elf")
+    set(CARGO_FEATURES_ARG "--features=std")
+endif()
+
+set(RUST_PROJECT_DIR "${CMAKE_CURRENT_LIST_DIR}")
+set(RUST_BUILD_DIR "${CMAKE_CURRENT_BINARY_DIR}")
+set(RUST_INCLUDE_DIR "${RUST_BUILD_DIR}")
+set(RUST_STATIC_LIBRARY "${RUST_BUILD_DIR}/${CARGO_TARGET}/${CARGO_BUILD_TYPE}/librustlib.a")
+
+ExternalProject_Add(
+    rustlib_project
+    PREFIX "${RUST_PROJECT_DIR}"
+    BUILD_COMMAND ${CMAKE_COMMAND} -E env
+        CARGO_BUILD_TARGET=${CARGO_TARGET}
+        CARGO_BUILD_TARGET_DIR=${RUST_BUILD_DIR}
+        cargo build ${CARGO_BUILD_ARG} ${CARGO_FEATURES_ARG}
+    DOWNLOAD_COMMAND ""
+    CONFIGURE_COMMAND ""
+    INSTALL_COMMAND ""
+    BUILD_ALWAYS TRUE
+    TMP_DIR "${RUST_BUILD_DIR}/tmp"
+    STAMP_DIR "${RUST_BUILD_DIR}/stamp"
+    DOWNLOAD_DIR "${RUST_BUILD_DIR}"
+    SOURCE_DIR "${RUST_PROJECT_DIR}"
+    BINARY_DIR "${RUST_PROJECT_DIR}"
+    INSTALL_DIR "${RUST_BUILD_DIR}"
+    BUILD_BYPRODUCTS "${RUST_STATIC_LIBRARY}"
```

Comment:
Is it appropriate to change this to something like:

```
BUILD_BYPRODUCTS "${RUST_BUILD_DIR}/${CARGO_BUILD_TYPE}" "${RUST_BUILD_DIR}/${CARGO_TARGET}" "${RUST_BUILD_DIR}/RustApi.h"
```

Right now Rust doesn't recompile after `idf.py` clean is run; it looks like `ExternalProjects` only clean the `BUILD_BYPRODUCTS`.

I'm not entirely sure which directories should be deleted: `release`, `xtensa-esp32-none-elf`, both?

Reply:
To save some searching, here are some relevant docs from [ExternalProject_Add](https://cmake.org/cmake/help/latest/module/ExternalProject.html#command:externalproject_add):

> BUILD_BYPRODUCTS
> These ultimately get passed through as BYPRODUCTS to the build step's own underlying call to add_custom_command().

and from one layer down, in [add_custom_command](https://cmake.org/cmake/help/latest/command/add_custom_command.html#command:add_custom_command):

> BYPRODUCTS
> The Makefile Generators will remove BYPRODUCTS and other GENERATED files during make clean.

Reply:
I'm going to add a `target` subdirectory to the component build directory so that all the Rust generated code is collected into a single directory.

Unfortunately, adding the target directory to the BUILD_BYPRODUCTS list won't work since the clean won't remove a non-empty directory.

However, it looks like I can repurpose the `config` command to run `cargo clean`. This way the Rust build files aren't removed during the the `idf.py clean` command, but a subsequent build will force the config step to clean and then fully rebuild the Rust project, which gives the desired behavior.

Reply:
Hmm - I thought checked the results of running `idf.py clean` and everything looked good. Not to doubt your results; I am on Mac and only remember trying Unix Makefiles.

Interesting idea about running `cargo clean`. Using cargo to clean Rust code seems like the least error-prone way to do things. In general it is a little quirky to clean on reconfigure, but in the context of `ExternalProject` that's a pretty good idea.